virttest/qemu_vm.py

```diff
@@ -2373,20 +2373,36 @@ def sort_key(dev):
                 if ats:
                     add_virtio_option("ats", ats, devices, device, dev_type)
 
-        # Add extra root_port at the end of the command line only if there is
-        # free slot on pci.0, discarding them otherwise
-        pcie_extra_root_port = int(params.get('pcie_extra_root_port', 0))
-        for num in range(pcie_extra_root_port):
-            try:
-                dev = devices.pcic_by_params(
-                    "pcie_extra_root_port_%s"
-                    % num, {"type": "pcie-root-port"})
-                devices.insert(dev)
-            except DeviceError:
-                logging.warning("No sufficient free slot for extra"
-                                " root port, discarding %d of them"
-                                % (pcie_extra_root_port - num))
-                break
+        if params.get('machine_type') == 'q35':
+            # Add extra root_port at the end of the command line only if there is
+            # free slot on pci.0, discarding them otherwise
+            port_index = 0
+            pcic_params = {'type': 'pcie-root-port'}
+            extra_port_num = int(params.get('pcie_extra_root_port', 0))
+            port_0_num = math.ceil(extra_port_num / 7)
+            for num in range(1, (port_0_num + 1)):
+                try:
+                    # enable multifunction for root port, and add a function
+                    # port to hotplug pcie device. And, keep the original
+                    # port name to avoid affecting exists test case.
+                    port_name = "pcie_extra_root_port_%d.0" % num
+                    root_port = devices.pcic_by_params(port_name, pcic_params)
+                    root_port.set_param('multifunction', 'on')
+                    devices.insert(root_port)
+                except DeviceError:
+                    logging.warning("No sufficient free slot for extra"
+                                    " root port, discarding %d of them"
+                                    % (extra_port_num - num))
+                    break
+                port_0_addr = root_port.get_param('addr')
+                func_port_num = 7 if num != port_0_num else (extra_port_num % 7)
```

Comment:
Would get at least one extra port than requested, for example, by set `pcie_extra_root_port = 1` it would get 2; by set `pcie_extra_root_port = 10`, it would get 12.

Please help check it, thanks.

Reply:
yes, if pcie_extra_root_port = 1,  two pcie-root-port will add,  port 0 (func0) set multifunction to 'on',  pcie device will be hotplug to func 1.
@jingzhao84  would you like to double confirm it.

Thanks,
Xu

Reply:
double confirm with @jingzhao84 , update code to make pcie device hotplug from func 0